fix: Import `moment` in `datetime.js` for web portal

marination opened the pull request:
## **Issue:**
> **TLDR: Web bundle needs moment imported in it, before datetime.js**
- On certain pages due to `moment` being undefined, it stops loading further js assets
- In this example, due to the error: frappe has no method `is_user_logged_in()` bound to it (`website.js not loaded due to error halting the process`)
- **Login** is visible even though the user is already logged in (works on some other pages)
   <img width="1419" alt="Screenshot 2022-02-02 at 10 23 27 AM" src="https://user-images.githubusercontent.com/25857446/152095413-cc0a147d-855c-45b5-84be-85851d5f3ef9.png">
- This breaks assets from other apps as well (erpnext js assets here)
- `moment` is imported via  > `libs.bundle.js` > which is used in **desk** views
- frappe's web bundle does not contain moment due to which, `moment is not undefined` breaks loading other js assets
- for some odd reason, importing moment in the web bundle does not work, so this is a temporary fix for that

## **Fix**:
- Add a new lib file `moment.js` which handles importing moment and binding it to the window
- Replaced all moment imports by this file
- Moment was imported and then momentTimezone (which eventually overwrites moment). Import only **momentTimezone** since **i think** it imports and uses **moment** internally.
- After fix:
  <img width="1329" alt="Screenshot 2022-02-02 at 10 30 16 AM" src="https://user-images.githubusercontent.com/25857446/152095736-97ee3664-5625-4f26-a64e-08eb445d8323.png">

## **(TIL) Today I Learned:**
- Imports in `frappe-web.bundle.js` execute sequentially. 
- So since `datetime.js` needed `moment` we tried importing and binding `moment` to the window before `datetime.js`, hoping it would work
- It still broke though, because the window binding [`window.moment = moment`] **will not execute until the entire set of imports in the bundle is completed**.
- `datetime.js` was imported before that, due to which the obvious missing moment
- Although during the imports if a file is loaded, since that file is being imported, the contents of that file will execute

So if we put some console.log in the **web bundle**, **datetime.js**  and **moment.js** (new file), this is how it will execute:
<img width="147" alt="Screenshot 2022-02-02 at 9 01 35 PM" src="https://user-images.githubusercontent.com/25857446/152184906-fc200540-5843-442a-ab8c-3346582b8c3b.png">
> **Bundle is logged last because it will execute console.log once the entire bundle finishes loading**

marination committed:
fix: Import moment in datetime js for web portal

- moment is imported via libs.bundle.js which is used in desk views
- frappe's web bundle does not contain moment due to which, `undefined moment` breaks loading other js assets
- for some odd reason, importing moment in the web bundle does not work, so this is a temporary fix for that

marination commented:
Breaks desk view loading 😞 , because `moment` is redefined.
Changing the statement to `import "moment/min/moment-with-locales.js";` works, but the portal again breaks

marination committed:
fix: Load `moment` and bind to window from new file

- Introduce new file in `/lib` to import moment and bind to window
- Use new lib file to load moment in web bundle and libs bundle (desk)

Co-authored-by: Suraj Shetty <surajshetty3416@gmail.com>

surajshetty3416 commented:
@marination let's also remove moment assets that we are loading in `base.html`... it is no longer required.

https://github.com/frappe/frappe/blob/develop/frappe/templates/base.html#L108-L109

marination commented:
@surajshetty3416 Also here ?
https://github.com/frappe/frappe/blob/develop/frappe/website/doctype/web_form/templates/web_form.html#L105-L106
https://github.com/frappe/frappe/blob/develop/frappe/website/doctype/web_form/templates/web_form.html#L97-L98

marination committed:
chore: Add `moment` to bundle instead of injection via script tag

- Wherever moment is imported in html, use dependent bundle to import it instead for consistency
- Remove moment import from `base.html` since it uses `frappe-web.bundle.js` anyway

marination commented:
@surajshetty3416 unrelated `test_contextual_user_permission` failing

surajshetty3416 commented:
The failing test looks unrelated. Merging.

marination commented:
@Mergifyio backport version-14-beta-pre-release
